[WIP] Search only replicas (scale to zero) with Reader/Writer Separation

prudhvigodithi opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description

While I refactor the code and add additional tests, I’m creating this PR to gather early feedback. I will share the testing results in the comments. Thanks!

- The primary goal is to allow users to designate an index as search-only allowing only to have the search only replicas running when enabled via an API call `_searchonly/enable` (can be disabled as `_searchonly/disable`).

- With `_searchonly/enable` for an index the process has Two-Phase Scale-Down with a temporary block  for the duration of the scale-down operation and then explicitly replace it with a permanent block once all prerequisites (e.g., shard sync, flush, metadata updates) have been met.

- Eliminates the need for users to manually invoke the `_remotestore/_restore` API to recover search-only replicas with `_searchonly/enable`, has automatic recovery of search-only replicas from the remote store during cluster recovery. Teh default behavior is still honored in normal conditions https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/remote-store/index/#restoring-from-a-backup.

- To Do (Work on the cluster health, coming from https://github.com/opensearch-project/OpenSearch/issues/15306#issuecomment-2489733095 add this similar implementation)

### Related Issues
https://github.com/opensearch-project/OpenSearch/issues/16720 and part of https://github.com/opensearch-project/OpenSearch/issues/15306

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

prudhvigodithi commented:
While I refactor the code and add additional tests, I’m creating this PR to gather early feedback please take a look and add your thoughts. I will share the testing results in the comments. Thanks!
@mch2 @shwetathareja @msfroh @getsaurabh02

prudhvigodithi commented:
## I went through and tested the following scenarios

### Scenario 1: Search-Only Replicas Recovery with Persistent Data Directory and when `cluster.remote_store.state.enabled` is set to false 

With the following settings, OpenSearch was started using:

```sh
./gradlew clean run -PnumNodes=6 --data-dir=/tmp/foo
```
<details><summary>OpenSearch settings</summary>
<p>
    <pre>
    <code>
setting 'path.repo', '["/tmp/my-repo"]'
setting 'opensearch.experimental.feature.read.write.split.enabled', 'true'
setting 'node.attr.remote_store.segment.repository', 'my-repository'
setting 'node.attr.remote_store.translog.repository', 'my-repository'
setting 'node.attr.remote_store.repository.my-repository.type', 'fs'
setting 'node.attr.remote_store.state.repository', 'my-repository'
setting 'node.attr.remote_store.repository.my-repository.settings.location', '/tmp/my-repo'
    </code>
  </pre>
</p>
</details>

Shard Allocation Before Recovery
```
curl -X GET "localhost:9200/_cat/shards/my-index?v&h=index,shard,prirep,state,unassigned.reason,node,searchOnly"

index    shard prirep state   unassigned.reason node
my-index 0     p      STARTED                   runTask-0
my-index 0     s      STARTED                   runTask-4
my-index 0     r      STARTED                   runTask-2
my-index 1     p      STARTED                   runTask-3
my-index 1     r      STARTED                   runTask-1
my-index 1     s      STARTED                   runTask-5
```
On restart (terminate the process) everything comes back as running. With search only enabled (`/_searchonly/enable`) after restart only search replicas are up as running and works as expected.
```
curl -X GET "localhost:9200/_cat/shards/my-index?v&h=index,shard,prirep,state,unassigned.reason,node,searchOnly"
index    shard prirep state   unassigned.reason node
my-index 0     s      STARTED                   runTask-2
my-index 1     s      STARTED                   runTask-1
``` 

### Scenario 2: No Data Directory Preservation and when `cluster.remote_store.state.enabled` is set t o false  – Index Lost After process Restart (Recovery)

In this scenario, OpenSearch is started without preserving the data directory, meaning that all local shard data is lost upon Recovery.

```
./gradlew clean run -PnumNodes=6
```
<details><summary>OpenSearch settings</summary>
<p>
    <pre>
    <code>
setting 'path.repo', '["/tmp/my-repo"]'
setting 'opensearch.experimental.feature.read.write.split.enabled', 'true'
setting 'node.attr.remote_store.segment.repository', 'my-repository'
setting 'node.attr.remote_store.translog.repository', 'my-repository'
setting 'node.attr.remote_store.repository.my-repository.type', 'fs'
setting 'node.attr.remote_store.state.repository', 'my-repository'
setting 'node.attr.remote_store.repository.my-repository.settings.location', '/tmp/my-repo'
    </code>
  </pre>
</p>
</details>

#### Behavior After Recovery:
- Upon terminating the process and restarting OpenSearch, the index is completely lost.
- Any attempt to retrieve the shard state results in an index not found exception.
```
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [my-index]","index":"my-index","resource.id":"my-index","resource.type":"index_or_alias","index_uuid":"_na_"}],"type":"index_not_found_exception","reason":"no such index [my-index]","index":"my-index","resource.id":"my-index","resource.type":"index_or_alias","index_uuid":"_na_"},"status":404}%
```
- Even With Remote Restore `_remotestore/_restore?restore_all_shards=true`, index remains unavailable.
- Even after recreating the index manually and attempting to restore, documents do not get picked up.
- Since during testing `--data-dir` was not used, local data (including cluster metadata) is wiped on recovery.
- Because the cluster state is lost, OpenSearch no longer has any reference to index.

### Scenario 3: Cluster Remote Store State Enabled (`cluster.remote_store.state.enabled` is set to true and with no persistent data directory) – Primary Shards Remain Unassigned After Recovery.

```
./gradlew clean run -PnumNodes=6
```
<details><summary>OpenSearch settings</summary>
<p>
    <pre>
    <code>
setting 'path.repo', '["/tmp/my-repo"]'
setting 'opensearch.experimental.feature.read.write.split.enabled', 'true'
setting 'node.attr.remote_store.segment.repository', 'my-repository'
setting 'node.attr.remote_store.translog.repository', 'my-repository'
setting 'node.attr.remote_store.repository.my-repository.type', 'fs'
setting 'node.attr.remote_store.state.repository', 'my-repository'
setting 'node.attr.remote_store.repository.my-repository.settings.location', '/tmp/my-repo'
setting 'cluster.remote_store.state.enabled', 'true'
    </code>
  </pre>
</p>
</details>


#### Shard Allocation After Recovery
```
curl -X GET "localhost:9200/_cat/shards/my-index?v&h=index,shard,prirep,state,unassigned.reason,node,searchOnly"
index    shard prirep state      unassigned.reason node
my-index 0     p      UNASSIGNED CLUSTER_RECOVERED 
my-index 0     s      UNASSIGNED CLUSTER_RECOVERED 
my-index 0     r      UNASSIGNED CLUSTER_RECOVERED 
my-index 1     p      UNASSIGNED CLUSTER_RECOVERED 
my-index 1     r      UNASSIGNED CLUSTER_RECOVERED 
my-index 1     s      UNASSIGNED CLUSTER_RECOVERED 
```
Issue: Primary Shards Remain Unassigned
Despite `cluster.remote_store.state.enabled` is true, the primary shards are not automatically assigned after restart ( replicating the recovery). The error message states:
```
"allocate_explanation": "cannot allocate because a previous copy of the primary shard existed but can no longer be found on the nodes in the cluster"
```
- Remote store only contains segments and translogs, NOT active shard copies.
- Since `--data-dir` was not used (data directory is not used), local copies of primary shards are lost.
- OpenSearch does not automatically restore primaries from the remote store without explicit intervention.
```
curl -X POST "http://localhost:9200/_remotestore/_restore" -H 'Content-Type: application/json' -d'  
{                
  "indices": ["my-index"]
}
'
```
- However with this PR, when `_searchonly` is enabled **Search-Only Replicas Recover Without a Primary**. Since `cluster.remote_store.state.enabled` is true, OpenSearch remembers the index exists after restart. The allocation logic skips checking for an active primary for search-only replicas.This allows search-only replicas to be assigned to a node, even without an existing primary. However without `_searchonly` the behavior is same for all replicas, wanted to give an advantage for users with `_searchonly` enabled indicies, for these indices they should not care `_remotestore/_restore` as we are not dealing with primaries.
   - Search-only replicas can recover automatically from the remote store.
   - Search queries remain functional.
   - Cluster state correctly remembers the index, but does not bring up primaries as `_searchonly` is enabled.

- The default behavior is OpenSearch does not assume lost primaries should be re-created from remote storage.It waits for explicit user intervention to restore primary shards (`_remotestore/_restore`). Is this by design ?

### Scenario 4: Persistent Data Directory with Remote Store State – Seamless Recovery of Primaries, replicas and search-only replicas

```
./gradlew clean run -PnumNodes=6 --data-dir=/tmp/foo
```

<details><summary>OpenSearch settings</summary>
<p>
    <pre>
    <code>
setting 'path.repo', '["/tmp/my-repo"]'
setting 'opensearch.experimental.feature.read.write.split.enabled', 'true'
setting 'node.attr.remote_store.segment.repository', 'my-repository'
setting 'node.attr.remote_store.translog.repository', 'my-repository'
setting 'node.attr.remote_store.repository.my-repository.type', 'fs'
setting 'node.attr.remote_store.state.repository', 'my-repository'
setting 'node.attr.remote_store.repository.my-repository.settings.location', '/tmp/my-repo'
setting 'cluster.remote_store.state.enabled', 'true'
    </code>
  </pre>
</p>
</details>


#### Upon recovery (no intervention is required )
```
curl -X GET "localhost:9200/_cat/shards/my-index?v&h=index,shard,prirep,state,unassigned.reason,node,searchOnly"
index    shard prirep state   unassigned.reason node
my-index 0     p      STARTED                   runTask-0
my-index 0     r      STARTED                   runTask-2
my-index 0     s      STARTED                   runTask-4
my-index 1     p      STARTED                   runTask-5
my-index 1     r      STARTED                   runTask-3
my-index 1     s      STARTED                   runTask-1
```
- All primary and replica shards successfully recover since the cluster metadata is retained in the persistent data directory.

If search-only mode is enabled to index, OpenSearch correctly brings up only search replicas while removing primary and regular replicas.

```
curl -X GET "localhost:9200/_cat/shards/my-index?v&h=index,shard,prirep,state,unassigned.reason,node,searchOnly"
index    shard prirep state   unassigned.reason node
my-index 0     s      STARTED                   runTask-3
my-index 1     s      STARTED                   runTask-3
```
- Only search replicas (SORs) are restored, as expected.

prudhvigodithi commented:
Coming from https://github.com/opensearch-project/OpenSearch/pull/17299#issuecomment-2643956842 @shwetathareja can you please go over scenario 2 and 3 and if it make sense. I wanted to understand  why `_remotestore/_restore` is required in these scenarios and I wanted to give advantage for users ti remove this intervention for search only indices.
Thanks 
@mch2

prudhvigodithi commented:
I have updated the PR to adjust the cluster health configuration using only search replicas and to incorporate the changes made when `_searchonly` is enabled, the change is not too big hence going with the same PR.

prudhvigodithi commented:
Adding @sachinpkale can you please take a look at this comment https://github.com/opensearch-project/OpenSearch/pull/17299#issuecomment-2643967803 and provide your thoughts to understand why `_remotestore/_restore` is required (Scenario 3 from https://github.com/opensearch-project/OpenSearch/pull/17299#issuecomment-2643956842) and why the cluster cannot be auto recovered, is there any strong reason for this manual intervention to run the API? 
```
curl -X GET "localhost:9200/_cat/shards/my-index?v&h=index,shard,prirep,state,unassigned.reason,node,searchOnly"
index    shard prirep state      unassigned.reason node
my-index 0     p      UNASSIGNED CLUSTER_RECOVERED 
my-index 0     s      UNASSIGNED CLUSTER_RECOVERED 
my-index 0     r      UNASSIGNED CLUSTER_RECOVERED 
my-index 1     p      UNASSIGNED CLUSTER_RECOVERED 
my-index 1     s      UNASSIGNED CLUSTER_RECOVERED 
my-index 1     r      UNASSIGNED CLUSTER_RECOVERED 
```

I dint get much info from the docs https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/remote-store/index/#restoring-from-a-backup.

prudhvigodithi commented on server/src/main/java/org/opensearch/rest/action/admin/indices/RestSearchonlyAction.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public List<Route> routes() {
+        return asList(new Route(POST, "/{index}/_searchonly/enable"), new Route(POST, "/{index}/_searchonly/disable"));
```
Thanks will take a look at this to go with a generic and which finds a wider applicability.

prudhvigodithi replied:
Initially I started with `_scale` https://github.com/opensearch-project/OpenSearch/issues/16720#issuecomment-2578385851. May be we can have ?
```
POST /{index}/_scale

{
  "search-only": true
}
```

Adding @msfroh @mch2 @getsaurabh02

shwetathareja replied:
as per original discussion @prudhvigodithi _scale is more intuitive

Bukhtawar replied:
Unfortunately you'll need to make a corresponding set of changes in the action names and probably the class names for consistency

prudhvigodithi replied:
Sure will do that. I can change to `RestScaleSearchOnlyAction`. Thanks

shwetathareja commented:
In continuation to above discussion, search-only replicas can be auto recovered from remote store without any caveats as it is read-only and not taking any active writes.

prudhvigodithi commented:
Thanks @shwetathareja this is more a primary durability concern and to avoid split-brain scenarios, if the original primary is merely partitioned and still accepting writes, automatically creating a new primary from remote store risks silent data loss. Hence, OpenSearch requires explicit user intervention before promoting a new primary. However, search replicas we can move forward with the idea to auto recover as they doesn't accept any writes. 

I will update the API code `_remotestore/_restore` to throw an error something like the following so that it will block the API execution and users and aware about the index in search only mode, this way the API wont interfere with search only mode.

```
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Skipping _remotestore/_restore for all selected indices as search-only mode is enabled."}],"type":"illegal_argument_exception","reason":"Skipping _remotestore/_restore for all selected indices as search-only mode is enabled."},"status":400}
```

Thanks
@mch2

Bukhtawar commented:
> Thanks @shwetathareja this is more a primary durability concern and to avoid split-brain scenarios, if the original primary is merely partitioned and still accepting writes, automatically creating a new primary from remote store risks silent data loss. Hence, OpenSearch requires explicit user intervention before promoting a new primary. However, search replicas we can move forward with the idea to auto recover as they doesn't accept any writes.
> 
> I will update the API code `_remotestore/_restore` to throw an error something like the following so that it will block the API execution and users and aware about the index in search only mode, this way the API wont interfere with search only mode.
> 
> ```
> {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Skipping _remotestore/_restore for all selected indices as search-only mode is enabled."}],"type":"illegal_argument_exception","reason":"Skipping _remotestore/_restore for all selected indices as search-only mode is enabled."},"status":400}
> ```
> 
> Thanks @mch2

Curious what should be the expectation when the search only replica goes down, shouldn't that be re-hydrated from remote by default?

prudhvigodithi commented:
Pushed a commit to fix the https://github.com/opensearch-project/OpenSearch/pull/17299#issuecomment-2675846489 codecov score.

prudhvigodithi commented:
I have tested this `search_only` mode with some large data set (by running this script https://github.com/prudhvigodithi/automations/blob/main/scripts/opensearch-large-data-ingest.sh) on a 10 node docker environment and works as expected. When `search_only` is set to true able to see only the the primaries are synced and removed leaving only search replicas,  also tested the docs count `_count` and ran some random search queries.

```
curl -X GET "localhost:9200/_cat/shards/my-index?v&h=index,shard,prirep,state,unassigned.reason,node,searchOnly"
index    shard prirep state   unassigned.reason node
my-index 0     s      STARTED                   opensearch-node10
my-index 0     s      STARTED                   opensearch-node6
my-index 1     s      STARTED                   opensearch-node9
my-index 1     s      STARTED                   opensearch-node4
my-index 2     s      STARTED                   opensearch-node10
my-index 2     s      STARTED                   opensearch-node9
```

Tested with specific custom routing logic (where one of the shards does not have the data) and logic works as expected.

```
curl -s "localhost:9200/_cat/shards/my-index?h=index,shard,prirep,state,docs,store,node"
my-index 0 p STARTED       0   230b runTask-3
my-index 0 s STARTED       0   230b runTask-6
my-index 0 s STARTED       0   230b runTask-0
my-index 0 r STARTED       0   230b runTask-9
my-index 1 p STARTED 2368392 39.3mb runTask-5
my-index 1 s STARTED 2368392 39.3mb runTask-4
my-index 1 s STARTED 2368392 39.3mb runTask-8
my-index 1 r STARTED 2368392 39.6mb runTask-7
my-index 2 p STARTED 1181608 19.1mb runTask-4
my-index 2 s STARTED 1181608 19.1mb runTask-6
my-index 2 r STARTED 1181608 19.1mb runTask-2
my-index 2 s STARTED 1181608 19.1mb runTask-1
```

Bukhtawar commented:
> > Curious what should be the expectation when the search only replica goes down, shouldn't that be re-hydrated from remote by default?
> 
> Hey @Bukhtawar
> 
> The search replicas start by `EmptyStoreRecoverySource.INSTANCE` during recovery, I have added a logic in `ReplicaShardAllocator` to find the available candidate node for allocation decision. Now by default with `EmptyStoreRecoverySource.INSTANCE` it load the complete set of segments that make up the shard from remote store and this repeats for every recovery scenario and dont need `¸`.
> 
> Now for default (when search_only is not enabled) the behavior is not changed and can be restored from `_remotestore/_restore` (with no persistent data directory) for primary and all replica types. From [#17299 (comment)](https://github.com/opensearch-project/OpenSearch/pull/17299#issuecomment-2643956842) Scenario 3.
> 
> Also we can always disable (brings back the initial state of the index) and enable the search_only which will reinitialize the replicas back again.
> 
> Adding @mch2 to provide any details If I'm missing.
> 
> Thanks

For `search_only` the ideal behaviour should be the cluster should never go `RED` in the first place so whatever is the behaviour with `_remotestore/_restore` on a non-RED cluster should be honoured to keep the API response consistent. While there are cases when the cluster can go RED too like allocation deciders hitting the max shard limit and preventing allocation of those shards.

prudhvigodithi commented:
> For `search_only` the ideal behaviour should be the cluster should never go `RED` in the first place so whatever is the behaviour with `_remotestore/_restore` on a non-RED cluster should be honoured to keep the API response consistent. While there are cases when the cluster can go RED too like allocation deciders hitting the max shard limit and preventing allocation of those shards.

With the current code when `search_only` mode is enabled the cluster is healthy i.e. it stays GREEN showing search replicas.
```
 curl -X GET "localhost:9200/_cat/shards/my-index?v&h=index,shard,prirep,state,unassigned.reason,node,searchOnly"
index    shard prirep state   unassigned.reason node
my-index 0     s      STARTED                   runTask-0
my-index 1     s      STARTED                   runTask-9
my-index 2     s      STARTED                   runTask-1

curl -X GET "localhost:9200/_cluster/health?pretty"  | jq '.'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   531  100   531    0     0   158k      0 --:--:-- --:--:-- --:--:--  172k
{
  "cluster_name": "runTask",
  "status": "green",
  "timed_out": false,
  "number_of_nodes": 10,
  "number_of_data_nodes": 10,
  "discovered_master": true,
  "discovered_cluster_manager": true,
  "active_primary_shards": 0,
  "active_shards": 3,
  "relocating_shards": 0,
  "initializing_shards": 0,
  "unassigned_shards": 0,
  "delayed_unassigned_shards": 0,
  "number_of_pending_tasks": 0,
  "number_of_in_flight_fetch": 0,
  "task_max_waiting_in_queue_millis": 0,
  "active_shards_percent_as_number": 100.0
}

curl -X GET "localhost:9200/_cluster/health?level=shards&pretty"
{
  "cluster_name" : "runTask",
  "status" : "green",
  "timed_out" : false,
  "number_of_nodes" : 10,
  "number_of_data_nodes" : 10,
  "discovered_master" : true,
  "discovered_cluster_manager" : true,
  "active_primary_shards" : 0,
  "active_shards" : 3,
  "relocating_shards" : 0,
  "initializing_shards" : 0,
  "unassigned_shards" : 0,
  "delayed_unassigned_shards" : 0,
  "number_of_pending_tasks" : 0,
  "number_of_in_flight_fetch" : 0,
  "task_max_waiting_in_queue_millis" : 0,
  "active_shards_percent_as_number" : 100.0,
  "indices" : {
    "my-index" : {
      "status" : "green",
      "number_of_shards" : 3,
      "number_of_replicas" : 1,
      "active_primary_shards" : 0,
      "active_shards" : 3,
      "relocating_shards" : 0,
      "initializing_shards" : 0,
      "unassigned_shards" : 0,
      "shards" : {
        "0" : {
          "status" : "green",
          "primary_active" : false,
          "active_shards" : 1,
          "relocating_shards" : 0,
          "initializing_shards" : 0,
          "unassigned_shards" : 0
        },
        "1" : {
          "status" : "green",
          "primary_active" : false,
          "active_shards" : 1,
          "relocating_shards" : 0,
          "initializing_shards" : 0,
          "unassigned_shards" : 0
        },
        "2" : {
          "status" : "green",
          "primary_active" : false,
          "active_shards" : 1,
          "relocating_shards" : 0,
          "initializing_shards" : 0,
          "unassigned_shards" : 0
        }
      }
    }
  }
}


```

Coming to the recovery for search replicas will work automatically using the empty store recovery mechanism. In this case, there’s no need to run the `_remotestore/_restore` API because the search replicas recover as expected without it. Meanwhile, if `search_only` is not enabled for an index, then the standard behavior is using `_remotestore/_restore` to recover both primary and replica shards (when the data dir is not persitent) which remains unchanged. 

@Bukhtawar is this what you are suggesting as well ?

prudhvigodithi commented on gradle/run.gradle:
```diff
@@ -45,6 +45,15 @@ testClusters {
+    setting 'node.attr.remote_store.translog.repository', 'my-repository'
+    setting 'node.attr.remote_store.repository.my-repository.type', 'fs'
+    setting 'node.attr.remote_store.state.repository', 'my-repository'
+    setting 'node.attr.remote_store.repository.my-repository.settings.location', '/tmp/my-repo'
```
I have used these settings to test the search only replicas with remote store as filesystem, will remove them and change to default once I have the green tests and finalized the PR review.

mch2 commented on server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java:
```diff
@@ -73,23 +73,30 @@ public void getCheckpointMetadata(
-                            null
+
+            if (mdFile == null) {
+                listener.onResponse(new CheckpointInfoResponse(indexShard.getLatestReplicationCheckpoint(), Collections.emptyMap(), null));
```
I think this change is so that SRs can become active and sync with remote before a primary has uploaded anything correct?  I don't think we want to make the same change for other replicas that rely on primary initialization to complete first.

prudhvigodithi replied:
I see your point let me handle a case for `isSearchOnly`. Something like ` } else if (indexShard.routingEntry().isSearchOnly()) {`.

prudhvigodithi commented on server/src/main/java/org/opensearch/cluster/routing/IndexRoutingTable.java:
```diff
@@ -425,8 +442,33 @@ public Builder initializeAsNew(IndexMetadata indexMetadata) {
          * Initializes an existing index.
          */
+
         public Builder initializeAsRecovery(IndexMetadata indexMetadata) {
```
I see `initializeEmpty` is generically called across those methods, let me update.

prudhvigodithi replied:
Now after thinking It's logical that these operations would need different behavior than recovery?. When reopening or importing an index, we should typically want all shards (primaries, regular replicas, and search replicas) to be initialized properly, not just search replicas, WDYT ?

prudhvigodithi replied:
Thinking back yes we can move this logic to common method `initializeEmpty`, but at this there is an issue with `_close` API when search replicas are enabled https://github.com/opensearch-project/OpenSearch/issues/15306#issuecomment-2686522304. For this PR I will move and keep the code in `initializeEmpty`.

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:
```diff
@@ -114,13 +120,29 @@ private void bootstrapIndexWithOutSearchReplicas(ReplicationType replicationType
+            future.actionGet();
+        });
+
+        assertTrue(exception.getMessage().contains("Cannot restore index [" + INDEX_NAME + "] because search-only mode is enabled"));
```
How about something like `Cannot use _remotestore/_restore to search_only mode enabled index [INDEX_NAME]` ?
Today for example we have 2 indices, in normal scenario the `_remotestore/_restore` errors out if one among them fails or has error
```
 curl -X POST "http://localhost:9200/_remotestore/_restore" -H 'Content-Type: application/json' -d'
{
  "indices": ["my_search_index", "my-index"]
}
'
```
Honoring the same if for example `my-index` is search_only mode enabled (irrespective of the order in the list) the entire operation fails.
```
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Cannot use _remotestore/_restore to search_only mode enabled index [my-index]."}],"type":"illegal_argument_exception","reason":"Cannot use _remotestore/_restore to search_only mode enabled index [my-index]."},"status":400}%
```

In before commit I have added a silent ignore, but with existing latest code it throws error stopping the operation.

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/SearchOnlyIT.java:
```diff
@@ -0,0 +1,549 @@
+        internalCluster().startDataOnlyNodes(6);
+
+        // Verify index is not found
+        expectThrows(IndexNotFoundException.class, () -> client().admin().indices().prepareGetIndex().setIndices(TEST_INDEX).get());
```
This is test coming from scenario 2 https://github.com/opensearch-project/OpenSearch/pull/17299#issuecomment-2643956842, its just to test the index does not come up, let me know your thoughts. This test to ensure the search replicas or any shards does not come up.

prudhvigodithi commented on server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:
```diff
@@ -397,6 +397,18 @@ protected void updateVisibleCheckpoint(long replicationId, IndexShard replicaSha
         ShardRouting primaryShard = clusterService.state().routingTable().shardRoutingTable(replicaShard.shardId()).primaryShard();
 
+        // Skip update if primary shard is null (which happens in search-only mode)
+        if (primaryShard == null) {
```
Coming from PR https://github.com/opensearch-project/OpenSearch/pull/17055/files, I can see some of the scale to zero tests broke, @vinaykpud can you please see if this change make sense ?

vinaykpud replied:
Its not related to the code you added here, Its because we change the behaviour of how replicas get the cached checkpoint of primary. Earlier if there is no checkpoint received at the replica it used to be `null` , but with my change we added default value as empty checkpoint. (https://github.com/vinaykpud/OpenSearch/blob/fcc57bf4265ca957b3452186e20810a6b547fcd4/server/src/main/java/org/opensearch/indices/replication/SegmentReplicator.java#L347)
So changing default to null should work.

prudhvigodithi replied:
Thanks I have removed it in my latest commit.

prudhvigodithi commented:
Looks like `org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT`  test suite is failing https://build.ci.opensearch.org/job/gradle-check/54033/testReport/.
```
expected [2xx] status code but api [search] returned [400 Bad Request] [{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Unknown NamedWriteable 
```
I havent seen this error before for this PR, @reta @andrross can you please confirm it does look like the failures are not from this PR.

cwperks commented:
Issue starting occurring after a backport to 2.x PR was merged: https://github.com/opensearch-project/OpenSearch/pull/17419

andrross commented:
@prudhvigodithi @cwperks I think you need to rebase the latest on main to get this commit to fix things: https://github.com/opensearch-project/OpenSearch/commit/7fc607ba031b2b687690cf647ce27e0adef3a8fb

prudhvigodithi commented:
> @prudhvigodithi @cwperks I think you need to rebase the latest on main to get this commit to fix things: [7fc607b](https://github.com/opensearch-project/OpenSearch/commit/7fc607ba031b2b687690cf647ce27e0adef3a8fb)

Yes seeing issues after the rebase and yes I can see the PR changes https://github.com/opensearch-project/OpenSearch/pull/17312 in my branch as well. I assume its the same case with your PR as well right https://github.com/opensearch-project/OpenSearch/pull/17389 ?

cwperks commented:
> @prudhvigodithi @cwperks I think you need to rebase the latest on main to get this commit to fix things: [7fc607b](https://github.com/opensearch-project/OpenSearch/commit/7fc607ba031b2b687690cf647ce27e0adef3a8fb)

Raise a PR to update the serialization of `execution_hint` from v_3_0_0 to v_2_19_1 since it was included in 2.19.1: https://github.com/opensearch-project/OpenSearch/pull/17492

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/SearchOnlyIT.java:
```diff
@@ -0,0 +1,549 @@
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class SearchOnlyIT extends RemoteStoreBaseIntegTestCase {
```
Sure will change this to `ScaleSearchOnlyIT`.

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/SearchOnlyIT.java:
```diff
@@ -0,0 +1,549 @@
+        assertEquals(1, shardTable.searchOnlyReplicas().stream().filter(ShardRouting::active).count());
+
+        // Scale down to search-only mode
+        assertAcked(client().admin().indices().prepareSearchOnly(TEST_INDEX).setSearchOnly(true).get());
```
How about `prepareScaleSearchOnly` ?

prudhvigodithi commented on server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -94,6 +94,7 @@
 import java.util.TreeSet;
 import java.util.function.Function;
 
+import static org.opensearch.action.admin.indices.scale.searchonly.TransportSearchOnlyAction.INDEX_SEARCHONLY_BLOCK;
```
Sure will do that

prudhvigodithi commented on server/src/main/java/org/opensearch/cluster/routing/IndexRoutingTable.java:
```diff
@@ -141,27 +144,41 @@ boolean validate(Metadata metadata) {
                         + "]"
                 );
+            } else if (routingNumberOfReplicas != expectedReplicas) {
+                // Just log if there's a mismatch but the index is search-only.
```
Ya just removed this.

prudhvigodithi commented on server/src/main/java/org/opensearch/cluster/routing/IndexRoutingTable.java:
```diff
@@ -94,6 +96,7 @@ public class IndexRoutingTable extends AbstractDiffable<IndexRoutingTable>
     private final Map<Integer, IndexShardRoutingTable> shards;
 
     private final List<ShardRouting> allActiveShards;
+    protected final Logger logger = LogManager.getLogger(this.getClass());
```
We dont need logger here.

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/SearchOnlyIT.java:
```diff
@@ -0,0 +1,549 @@
+
+        // Verify we can still search
+        SearchResponse searchResponse = client().prepareSearch(TEST_INDEX).get();
+        assertHitCount(searchResponse, 10);
```
Adding should help right ?
```
        assertBusy(() -> {
            SearchResponse searchResponse = client().prepareSearch(TEST_INDEX).get();
            assertHitCount(searchResponse, 10);
        });
  ```
 For scale API when search_only is enabled the ack'd is done after the final sync is completed and primaries and replicas are scaled down.

prudhvigodithi replied:
Or something like 
```
        for (int i = 0; i < 10; i++) {
            IndexResponse indexResponse = client().prepareIndex(TEST_INDEX)
                .setId(Integer.toString(i))
                .setSource("field1", "value" + i)
                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
                .get();
            assertEquals(RestStatus.CREATED, indexResponse.status());
        }

        assertBusy(() -> {
            SearchResponse searchResponse = client().prepareSearch(TEST_INDEX).get();
            assertHitCount(searchResponse, 10);
        }, 30, TimeUnit.SECONDS);
```
and after scale down something like
```
            SearchResponse searchResponse = client().prepareSearch(TEST_INDEX).get();
            assertHitCount(searchResponse, 10);
 ```

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/SearchOnlyIT.java:
```diff
@@ -0,0 +1,549 @@
+        ensureGreen(TEST_INDEX);
+
+        // Wait for shards to be properly allocated
+        assertBusy(() -> {
```
I see you mean to say if `ensureGreen(TEST_INDEX);` already checks this I dont need to re-validate the shard numbers again, let me check on that.

peterzhuamazon commented:
Seems like still flaky as I saw one run pass here on Ubuntu2404:
https://build.ci.opensearch.org/job/gradle-check/54280/console

Thanks.

prudhvigodithi commented:
> Seems like still flaky as I saw one run pass here on Ubuntu2404: https://build.ci.opensearch.org/job/gradle-check/54280/console

Yes its genuine test failure Peter.

peterzhuamazon commented:
Can try rebase:
* https://github.com/opensearch-project/OpenSearch/pull/17535
* https://github.com/opensearch-project/OpenSearch/pull/17546

prudhvigodithi commented:
> Can try rebase:

Just did Peter, but looks like the failures are coming from flaky tests, I can see the `AUTOCUT` automation linked this PR to the open issues

prudhvigodithi commented:
I have got a review from @mch2, tagging @shwetathareja @Bukhtawar @msfroh @andrross for 2nd review on this PR.
Thanks

shwetathareja commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/TransportSearchOnlyAction.java:
```diff
@@ -0,0 +1,458 @@
+
+    ShardSearchOnlyResponse syncSingleShard(IndexShard shard) throws Exception {
+        logger.info("Performing final sync and flush for shard {}", shard.shardId());
+        shard.sync();
+        shard.flush(new FlushRequest().force(true).waitIfOngoing(true));
+
+        if (shard.translogStats().getUncommittedOperations() > 0) {
+            String errorMsg = String.format(
+                Locale.ROOT,
+                "Shard [%s] still has %d uncommitted operations after flush. Please wait and retry the scale down operation.",
+                shard.shardId(),
+                shard.translogStats().getUncommittedOperations()
+            );
+            throw new IllegalStateException(errorMsg);
+        }
+
+        shard.waitForRemoteStoreSync();
```
What if primary is still receiving active traffic. Search-only block can prevent traffic if the data node hosting primary copy has applied the cluster state change successfully. In case, it didnt it will continue to take writes. In between translog can be zero as well, it will be misleading. You will need to put an explicit lock in translog so that no new further writes can be entertained

prudhvigodithi replied:
So the logic is:

if the request indicates a scale‐down, which submits a cluster state update to add a temporary block on the target indices. This temporary block prevents writes to the index during the preparation phase of scaling down

Only after the temporary block is in place then the remote shard sync requests sent which also adds a permanent search-only block. This has the Two-Phase Approach.

Having this do you think we still need a explicit lock mechanism in translog?

mch2 replied:
Wondering if we need to wrap this in `acquireAllPrimaryOperationsPermits` to resolve case locally similar to verify shard before close / verify block apis?

Can we test this in an IT that concurrently writes?

prudhvigodithi replied:
I have updated the `testFullSearchOnlyReplicasFullLifecycle` test method (part of `ScaleIndexIT`) to assert the writer blocks during the scale down operation.

prudhvigodithi replied:
Let me take a look at `acquireAllPrimaryOperationsPermits`.

prudhvigodithi replied:
I can see the current implementation already provides some safeguards:
- 1st add a temporary write block via cluster state update that prevents new operations.
- Verify that the block is in place before proceeding.
- Explicitly sync and flush each shard (in-flight operations).
- Verify there are no uncommitted operations after the flush.
- Then add a permanent search-only block only after successful verification.

prudhvigodithi replied:
Testing on an index with continuous traffic, even though we expect to perform search-only operations on a lower-traffic, idle index.

```
for i in {1..10000}; do 
  curl -X POST "http://localhost:9200/my-index/_doc/$i" \
    -H "Content-Type: application/json" \
    -d '{"domainAttributes":{"firmId":"'"$i"'"}}';
  echo "Indexing $i"
done
```

Showed error (uncommitted operations in translog )with in flight data
```
"error":{"root_cause":[{"type":"illegal_state_exception","reason":"Shard [[my-index][1]] still has 2 uncommitted operations after flush. Please wait and retry the scale down operation."}],"type":"illegal_state_exception","reason":"Shard [[my-index][1]] still has 2 uncommitted operations after flush. Please wait and retry the scale down operation."},"status":500}%                                                    
```

Thanks to @mch2 suggestions here using `acquireAllPrimaryOperationsPermits` solved the problem, updated the code to use `acquireAllPrimaryOperationsPermits` (updated the tests accordingly) in my latest commit.

Now with this change we should be able to convert an index to search_only on an write heavy index, where new requests once the block is applied now fails with 

```
{"error":{"root_cause":[{"type":"cluster_block_exception","reason":"index [my-index] blocked by: [FORBIDDEN/20/index preparing to scale down];"}],"type":"cluster_block_exception","reason":"index [my-index] blocked by: [FORBIDDEN/20/index preparing to scale down];"},"status":403
```

prudhvigodithi replied:
Adding @shwetathareja

prudhvigodithi commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/TransportSearchOnlyAction.java:
```diff
@@ -0,0 +1,458 @@
+
+        @Override
+        public void onFailure(String source, Exception e) {
+            logger.error("Failed to finalize scale-down operation", e);
```
Hey @shwetathareja getting back to this PR, my idea since the index has to be readonly, during scale down if there is any error the main scale down (removing the primaries and regular replicas) wont happen, but the block (search_only) persist, this way there wont new any new writes (even accidental) to index and the user can always re-try the scale down (`"search_only": true`) again, the `"search_only": false` will anyhow remove the lock.

prudhvigodithi commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/NodeSearchOnlyRequest.java:
```diff
@@ -0,0 +1,87 @@
+ * ensuring all operations are committed and the remote store is synced. This is a crucial step in
+ * the scale-down process to ensure no data loss occurs when the index transitions to search-only mode.
+ */
+class NodeSearchOnlyRequest extends TransportRequest {
```
Sure, I will refactor to remove the `SearchOnly` prefix.

prudhvigodithi replied:
> On a different note, SearchOnly prefix is not very reader friendly

@shwetathareja I have refactored the code and all the class names better to denote `Scale` and remove the `SearchOnly`.

prudhvigodithi commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/TransportSearchOnlyAction.java:
```diff
@@ -0,0 +1,458 @@
+            }
+
+            shardResponses.add(syncSingleShard(shard));
+        }
```
This is new to me, but seeing other existing code I have used `ThreadPool.Names.MANAGEMENT` (`threadPool.executor(SHARD_SYNC_EXECUTOR).execute(()`) for this process so that it does not block the other threads, can you please check?

prudhvigodithi commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/TransportSearchOnlyAction.java:
```diff
@@ -0,0 +1,458 @@
+ * The scale operation is implemented as a series of cluster state update tasks to ensure
+ * atomicity and consistency throughout the transition.
+ */
+public class TransportSearchOnlyAction extends TransportClusterManagerNodeAction<SearchOnlyRequest, AcknowledgedResponse> {
```
The `TransportScalingAction` is now `TransportScaleIndexAction`. Anyhow I have the ClusterStateUpdateTasks as inner classes, do you want me to move them outside?

prudhvigodithi commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/ShardSearchOnlyResponse.java:
```diff
@@ -0,0 +1,109 @@
+ * The cluster manager uses this information from all primary shards to decide
+ * whether it's safe to finalize a scale-down operation.
+ */
+class ShardSearchOnlyResponse implements Writeable {
```
The `SearchOnlyShardResponse` is now `ScaleIndexShardResponse` and in sync with the other package classes.

prudhvigodithi commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/NodeSearchOnlyResponse.java:
```diff
@@ -0,0 +1,89 @@
+ * additional synchronization, which would indicate the scale operation should be delayed until
+ * the cluster reaches a stable state.
+ */
+class NodeSearchOnlyResponse extends TransportResponse {
```
The `SearchOnlyShardNodeResponse` is now `ScaleIndexNodeResponse`.

prudhvigodithi commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/NodeSearchOnlyRequest.java:
```diff
@@ -0,0 +1,87 @@
+ * the scale-down process to ensure no data loss occurs when the index transitions to search-only mode.
+ */
+class NodeSearchOnlyRequest extends TransportRequest {
+    private final String index;
```
Coming from the other request classes `DeleteIndexRequest`, `OpenIndexRequest`, `CreateIndexRequest`, `CloseIndexRequest` we use either String[] or String, hence I used as String. Let me double check on using as `Index`.

prudhvigodithi commented:
Coming from [Metrics Dashboard](https://metrics.opensearch.org/_dashboards/app/dashboards?security_tenant=global#/view/e5e64d40-ed31-11ee-be99-69d1dbc75083?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-90d,to:now))&_a=(description:'',filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'2e729590-eb26-11ee-b008-6583d3dd450c',key:pull_request,negate:!f,params:(query:'17299'),type:phrase),query:(match_phrase:(pull_request:'17299')))),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!t,title:'OpenSearch%20Gradle%20Check%20Report',viewMode:view)) the failures are all related to the flaky tests :(

msfroh commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexAction.java:
```diff
@@ -0,0 +1,47 @@
+ * The action is registered with the name "indices:admin/scale/search_only" in the
+ * transport action registry.
+ */
+public class ScaleIndexAction extends ActionType<AcknowledgedResponse> {
```
I'm not a big fan of this action name. 

It feels too generic for the case that it's trying to address. I'm worried that if we go for the specific name, we'll end up in a similar situation to our existing "search" node type, which really means "searchable snapshot".

That said, the API part of it is the `NAME` field, which does include `search_only`, which is good. For the class name, we have some flexibility to change in the future.

prudhvigodithi replied:
Thanks @msfroh thats a good point, I though about it and hence updated the `NAME` and package to include `searchonly`.  So the `ScaleIndexAction` is inside the searchonly package.

msfroh commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexOperationValidator.java:
```diff
@@ -0,0 +1,87 @@
+                if (indexMetadata.getNumberOfSearchOnlyReplicas() == 0) {
+                    throw new IllegalArgumentException("Cannot scale to zero without search replicas for index: " + index);
+                }
+                if (!indexMetadata.getSettings().getAsBoolean(IndexMetadata.SETTING_REMOTE_STORE_ENABLED, false)) {
```
Nit: There's a code convention that we inherited from Lucene where instead of a leading `!`, it's a good idea to do `== false`. It makes it more visible that the condition applies if the value is false. I don't have very strong feelings on the convention, but I've adopted it.

That applies to the segrep condition below as well.

prudhvigodithi replied:
Sure Froh will update to use `== false`.

msfroh commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexRequestBuilder.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Constructs a new builder for scaling an index.
+     * <p>
+     * By default, the operation is a scale-up operation (from search-only to normal mode).
```
Does it make sense to have a default? Or should we force clients to decide which operation they want?

I think we only really need the three-arg constructor, which then means we can drop the setter.

prudhvigodithi replied:
Coming from https://github.com/opensearch-project/OpenSearch/pull/17299#discussion_r2015156859 let me remove defaults and add three-arg constructor.

msfroh commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexResponse.java:
```diff
@@ -0,0 +1,137 @@
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+        builder.startObject();
+        builder.field("has_failures", hasFailures);
```
Do we need the `has_failures` property? Or can we just rely on `failure_reason` being non-null?

prudhvigodithi replied:
So the relation is `has_failures` is true if and only if `failure_reason` is present, coming from the method `buildFailureReason` part of the same class.

prudhvigodithi replied:
Yes we can remove the `builder.field("has_failures", hasFailures);`.

msfroh commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexShardSyncManager.java:
```diff
@@ -0,0 +1,217 @@
+     * @param responses the collection of responses from all nodes
+     * @param listener  the listener to notify with the aggregated result
+     */
+    void validateNodeResponses(Collection<ScaleIndexNodeResponse> responses, ActionListener<ScaleIndexResponse> listener) {
```
Is this method used in the production code? It looks like it's only called from tests.

prudhvigodithi replied:
Yes Froh it is used.

In the `handleShardSyncResponses` method (part of `TransportScaleIndexAction`), the `validateNodeResponses`  is called to verify that all shards are in a consistent state before proceeding with the scale-down operation. 
The `validateNodeResponses` method performs some critical validations. It iterates through all shard responses from all nodes and checks if any shard has uncommitted operations and needs synchronization If any shards fail these checks, it collects their IDs and generates a detailed error message.

msfroh replied:
Got it, thanks! The diff for `TransportScaleIndexAction` wasn't fetched, so I couldn't find where it was used. It makes sense.

msfroh commented on server/src/main/java/org/opensearch/transport/client/support/AbstractClient.java:
```diff
@@ -2145,6 +2146,11 @@ public void updateView(CreateViewAction.Request request, ActionListener<GetViewA
             return execute(UpdateViewAction.INSTANCE, request);
         }
+
+        public ScaleIndexRequestBuilder prepareScaleSearchOnly(String index) {
```
Related to my comment above about the `ScaleIndexRequestBuilder` and whether there should be a default, I'm more convinced by this method that we should force clients to select the direction of scaling. 

Right now, someone would call `prepareScaleSearchOnly` and get a `ScaleIndexRequestBuilder` that **scales up  from search only**, by default.

prudhvigodithi replied:
Sure Froh, let me update as `prepareScaleSearchOnly(String index, boolean searchOnly)` and so the `ScaleIndexRequestBuilder(OpenSearchClient client, boolean searchOnly, String index), this was we can have three-arg constructor and remove the defaults.

prudhvigodithi commented:
Thanks for your inputs @msfroh, I have addressed your comments and can you please take a look at the PR again ?
@mch2

mch2 commented on server/src/main/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexOperationValidator.java:
```diff
@@ -0,0 +1,88 @@
+                if (indexMetadata.getNumberOfSearchOnlyReplicas() == 0) {
+                    throw new IllegalArgumentException("Cannot scale to zero without search replicas for index: " + index);
+                }
+                if (indexMetadata.getSettings().getAsBoolean(IndexMetadata.SETTING_REMOTE_STORE_ENABLED, false) == false) {
```
If you pass/fetch the IndexSettings here, we already have methods already to validate these remote and segrep settings.

prudhvigodithi replied:
I see, anyhow we need the metadata to check the blocks, is it fine we go with this ?

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexIT.java:
```diff
@@ -0,0 +1,551 @@
+            assertHitCount(searchResponse, 10);
+        }, 30, TimeUnit.SECONDS);
+
+        assertBusy(() -> {
```
Got it let me remove this, I remember I have cleaned up the tests that count the shards and assert the number which is not needed with `ensureGreen`. Let me see if I have missed more.

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexIT.java:
```diff
@@ -0,0 +1,551 @@
+    /**
+     * Tests restoring an index from search-only mode back to normal mode.
+     */
+    public void testScaleUpFromSearchOnly() throws Exception {
```
Sure let me combine and add a new test `testScaleDownAndUpWithSearchOnly`

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexIT.java:
```diff
@@ -0,0 +1,551 @@
+        );
+
+        assertTrue(
+            "Expected error about missing search replicas",
```
This test asserts that one cannot scale down to `search_only` when the the search replicas are not even configured, meaning for indices without `number_of_search_only_replicas`.

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexIT.java:
```diff
@@ -0,0 +1,551 @@
+     * Scenario 2: Tests behavior without data directory preservation
+     * and cluster.remote_store.state.enabled=false
+     */
+    public void testRecoveryWithoutDataDirPreservation() throws Exception {
```
Right we should get the `IndexNotFoundException`, coming from https://github.com/opensearch-project/OpenSearch/pull/17299#issuecomment-2643956842 scenario 2.

prudhvigodithi commented on server/src/internalClusterTest/java/org/opensearch/action/admin/indices/scale/searchonly/ScaleIndexIT.java:
```diff
@@ -0,0 +1,551 @@
+    /**
+    * Tests the ability to scale search replicas up and down while an index is in search-only mode.
+    */
+    public void testScaleSearchReplicasInSearchOnlyMode() throws Exception {
```
Sure let me add the full lifecycle to a single test.

prudhvigodithi commented:
Hey @mch2 I have updated the IT tests with your suggestions, can you please take a look at the PR again?
Thanks

prudhvigodithi commented:
After addressing the comments I have fixed the conflicts and updated the `CHANGELOG.md`, @msfroh and @mch2 can you please take a look, so that we can get this merged and battle tested for 3.0.0 beta release.
Thanks

mch2 commented on server/src/main/java/org/opensearch/indices/replication/SegmentReplicator.java:
```diff
@@ -345,7 +345,7 @@ SegmentReplicationTarget get(ShardId shardId) {
 
     ReplicationCheckpoint getPrimaryCheckpoint(ShardId shardId) {
-        return primaryCheckpoint.getOrDefault(shardId, ReplicationCheckpoint.empty(shardId));
+        return primaryCheckpoint.get(shardId);
```
Where is returning an empty checkpoint causing issues that this needs to be updated?

vinaykpud replied:
Earlier if there is no checkpoint received at the replica it used return `null` when `getPrimaryCheckpoint` called. But with  but with my change in this PR(In https://github.com/opensearch-project/OpenSearch/pull/17055/) 
we added default value ie empty checkpoint. So it broke some tests in scale to zero.

mch2 commented on server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java:
```diff
@@ -67,13 +67,32 @@ public void getCheckpointMetadata(
+                } else {
+                    // Regular replicas should not be active without metadata
+                    listener.onFailure(
+                        new IllegalStateException("Remote metadata file can't be null for active regular replica: " + indexShard.shardId())
```
I'm ok changing the assert to a failure case here as in prod this would be an NPE, but lets preserve the message.
```suggestion
                        new IllegalStateException("Remote metadata file can't be null if shard is active: " + indexShard.shardId())
```

mch2 commented on server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java:
```diff
@@ -67,13 +67,32 @@ public void getCheckpointMetadata(
+                    // Allow search-only replicas to become active without metadata
+                    logger.debug("Search-only replica proceeding without remote metadata: {}", indexShard.shardId());
+                    listener.onResponse(
+                        new CheckpointInfoResponse(indexShard.getLatestReplicationCheckpoint(), Collections.emptyMap(), null)
```
Lets note here that null infosBytes is safely handled by the caller `SegmentReplicationTarget`.

mch2 commented on server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java:
```diff
@@ -142,17 +141,14 @@ public RemoteRestoreResult restore(
-        boolean metadataFromRemoteStore = (restoreClusterUUID == null
-            || restoreClusterUUID.isEmpty()
-            || restoreClusterUUID.isBlank()) == false;
+        boolean metadataFromRemoteStore = restoreClusterUUID != null && !restoreClusterUUID.isEmpty();
```
This is not logically equivalent as isBlank will check for whitespace.  Why do we need to update this check?

prudhvigodithi replied:
I see, I dint expect we can get the restoreClusterUUID as whitespace and mess the boolean, while understanding the logic I modified this to do some clean up. But seeing your comment I will revert to default.
```
        boolean metadataFromRemoteStore = (restoreClusterUUID == null
            || restoreClusterUUID.isEmpty()
            || restoreClusterUUID.isBlank()) == false;
 ```

mch2 replied:
yes pls :) This change is already really large please only change whats needed.

mch2 commented on server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -248,14 +249,47 @@ protected AllocateUnassignedDecision getAllocationDecision(
+                // If there are no data nodes available, delay allocation.
+                return AllocateUnassignedDecision.delayed(0L, 0L, null);
+            } else {
+                // For non-search-only replicas, if there is no active primary we do not attempt an allocation.
```
lets include the assertion that was removed above - 
```
            assert explain : "primary should only be null here if we are in explain mode, so we didn't "
                + "exit early when canBeAllocatedToAtLeastOneNode didn't return a YES decision";
```

mch2 commented on server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java:
```diff
@@ -248,14 +249,47 @@ protected AllocateUnassignedDecision getAllocationDecision(
+                .getAsBoolean(IndexMetadata.INDEX_BLOCKS_SEARCH_ONLY_SETTING.getKey(), false);
+
+            // Only proceed if both the shard is marked search-only and the index setting is enabled.
+            if (unassignedShard.isSearchOnly() && isIndexSearchOnly) {
```
Lets look up the index setting only if the shard is searchOnly? - 

```java
if (unassignedShard.isSearchOnly) {
  return getSearchReplicaAllocationDecision(....);
}
```

prudhvigodithi replied:
Let me created `getSearchReplicaAllocationDecision` and move the logic there and also move the `isIndexSearchOnly` inside the `if (unassignedShard.isSearchOnly()`.

mch2 commented on server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -166,6 +166,24 @@ public class IndexMetadata implements Diffable<IndexMetadata>, ToXContentFragmen
     );
 
+    // Block ID and block for scale operations (IDs 20-29 reserved for scaling)
+    public static final int INDEX_SEARCHONLY_BLOCK_ID = 20;
```
```suggestion
    public static final int INDEX_SEARCH_ONLY_BLOCK_ID = 20;
```

mch2 commented on server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -166,6 +166,24 @@ public class IndexMetadata implements Diffable<IndexMetadata>, ToXContentFragmen
+     * <p>
+     * This block prevents write operations to the index while allowing read operations.
+     */
+    public static final ClusterBlock INDEX_SEARCHONLY_BLOCK = new ClusterBlock(
```
```suggestion
    public static final ClusterBlock INDEX_SEARCH_ONLY_BLOCK = new ClusterBlock(
```

mch2 commented on server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:
```diff
@@ -93,6 +93,17 @@ public ClusterStateHealth(final ClusterState clusterState, final String[] concre
         indices = new HashMap<>();
+
+        boolean hasSearchOnlyIndices = false;
+        for (String index : concreteIndices) {
```
we already iterate over concreteIndices below on line 96, lets set this flag from there?

mch2 commented on server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:
```diff
@@ -123,6 +133,34 @@ public ClusterStateHealth(final ClusterState clusterState, final String[] concre
+                IndexMetadata indexMetadata = clusterState.metadata().index(index);
+                if (indexMetadata == null) continue;
+
+                boolean isSearchOnlyEnabled = indexMetadata.getSettings()
```
```suggestion
                boolean isSearchOnlyCluserBlockEnabled = indexMetadata.getSettings()
```

I think here and elsewhere we need to be clear to differentiate between "The index is search only", ie has search only shards and the index has the cluster block applied.

prudhvigodithi replied:
Sure let me find and update to `isSearchOnlyCluserBlockEnabled` (during the setting check `INDEX_BLOCKS_SEARCH_ONLY_SETTING`.)

prudhvigodithi replied:
I have updated this in my latest commit, please check.

mch2 commented on server/src/main/java/org/opensearch/cluster/health/ClusterStateHealth.java:
```diff
@@ -162,6 +200,16 @@ public ClusterStateHealth(
         boolean isIndexOrShardLevelHealthRequired = healthLevel == ClusterHealthRequest.Level.INDICES
             || healthLevel == ClusterHealthRequest.Level.SHARDS;
 
+        boolean hasSearchOnlyIndices = false;
```
pull this to a private function if repeated pls.

prudhvigodithi replied:
I have updated this in my latest commit, please check.

mch2 commented on server/src/main/java/org/opensearch/cluster/health/ClusterShardHealth.java:
```diff
@@ -229,9 +230,36 @@ public void writeTo(final StreamOutput out) throws IOException {
+        final int totalShards,
+        final IndexMetadata indexMetadata
+    ) {
+        boolean isSearchOnlyEnabled = indexMetadata.getSettings()
```
pls inline this after checking for null routing.

prudhvigodithi replied:
I have updated this in my latest commit, please check.

mch2 commented on server/src/main/java/org/opensearch/cluster/health/ClusterShardHealth.java:
```diff
@@ -113,7 +114,7 @@ public final class ClusterShardHealth implements Writeable, ToXContentFragment {
     private final boolean primaryActive;
 
-    public ClusterShardHealth(final int shardId, final IndexShardRoutingTable shardRoutingTable) {
+    public ClusterShardHealth(final int shardId, final IndexShardRoutingTable shardRoutingTable, final IndexMetadata indexMetadata) {
```
Similar to comment on getShardHealth, If all we need to know is if the cluster block is applied.  Lets overload the ctor here with a boolean and default it to false?

prudhvigodithi replied:
I have updated this in my latest commit, please check.

mch2 commented on server/src/main/java/org/opensearch/cluster/health/ClusterShardHealth.java:
```diff
@@ -229,9 +230,36 @@ public void writeTo(final StreamOutput out) throws IOException {
+        final ShardRouting primaryRouting,
+        final int activeShards,
+        final int totalShards,
+        final IndexMetadata indexMetadata
```
lets overload this and have the caller specify if null primary is allowed rather than passing metadata.

prudhvigodithi replied:
I have updated this in my latest commit, please check.

mch2 commented:
I think this is in a good state, I'd like to get this merged so we can bake this and iterate if needed before 3.0 cutoff.
@shwetathareja @Bukhtawar @msfroh Wondering if any of you would like to make another pass here, will hold for another day given the size of this.
